gpytorch/lazy/diag_lazy_tensor.py

```diff
@@ -3,159 +3,131 @@
+            res = res.expand(batch_shape + res.shape[-1:])
         return (res,)
 
     def _size(self):
-        if self._diag.ndimension() == 2:
-            return self._diag.size(0), self._diag.size(-1), self._diag.size(-1)
-        else:
-            return self._diag.size(-1), self._diag.size(-1)
+        return self._diag.shape + self._diag.shape[-1:]
+
+    def _t_matmul(self, rhs):
+        # Diagonal matrices always commute
+        return self._matmul(rhs)
 
     def _transpose_nonbatch(self):
         return self
 
-    def _batch_get_indices(self, batch_indices, left_indices, right_indices):
-        equal_indices = left_indices.eq(right_indices).type_as(self._diag)
-        return self._diag[batch_indices, left_indices] * equal_indices
-
-    def _get_indices(self, left_indices, right_indices):
-        equal_indices = left_indices.eq(right_indices).type_as(self._diag)
-        return self._diag[left_indices] * equal_indices
+    def abs(self):
+        return DiagLazyTensor(self._diag.abs())
 
     def add_diag(self, added_diag):
-        return DiagLazyTensor(self._diag + added_diag.expand_as(self._diag))
-
-    def __add__(self, other):
-        from .added_diag_lazy_tensor import AddedDiagLazyTensor
-        if isinstance(other, DiagLazyTensor):
-            return DiagLazyTensor(self._diag + other._diag)
-        else:
-            return AddedDiagLazyTensor(other, self)
+        return DiagLazyTensor(self._diag + added_diag)
 
     def diag(self):
         return self._diag
 
     def evaluate(self):
-        if self.ndimension() == 2:
-            return self._diag.diag()
-        else:
-            return super(DiagLazyTensor, self).evaluate()
+        return self._diag.unsqueeze(-1) * torch.eye(self._diag.shape[-1])
 
-    def sum_batch(self, sum_batch_size=None):
-        if sum_batch_size is None:
-            diag = self._diag.view(-1, self._diag.size(-1))
-        else:
-            diag = self._diag.view(-1, sum_batch_size, self._diag.size(-1))
+    def exp(self):
+        return DiagLazyTensor(self._diag.exp())
 
-        return self.__class__(diag.sum(-2))
+    def inverse(self):
+        return DiagLazyTensor(self._diag.reciprocal())
 
     def inv_matmul(self, tensor):
-        is_vec = False
-        if (self.dim() == 2 and tensor.dim() == 1):
-            tensor = tensor.unsqueeze(-1)
-            is_vec = True
-
-            if self.shape[-1] != tensor.numel():
-                raise RuntimeError(
-                    "LazyTensor (size={}) cannot be multiplied with right-hand-side Tensor (size={}).".format(
-                        self.shape, tensor.shape
-                    )
-                )
-
-        res = tensor.div(self._diag.unsqueeze(-1))
-        if is_vec:
-            res = res.squeeze(-1)
-        return res
+        return self.inverse()._matmul(tensor)
 
     def inv_quad_log_det(self, inv_quad_rhs=None, log_det=False, reduce_inv_quad=True):
-        if inv_quad_rhs is not None:
```

Comment:
I think all these checks are still necessary, and should work for multi-batch mode?